walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -47,16 +47,36 @@
     }
 
     // Add optional guests
-    // TODO: Use stream to make a deep copy
-    List<TimeRange> optionalMeetingTimes = new ArrayList<TimeRange>();
-    for (int i = 0; i < meetingTimes.size(); i++) {
-      optionalMeetingTimes.add(meetingTimes.get(i).fromStartDuration(meetingTimes.get(i).start(), 
-          meetingTimes.get(i).duration()));
-    }
-    optionalMeetingTimes = findSchedule(possibleTimes, events, optionalGuests, request);
-    if (!optionalMeetingTimes.isEmpty() || attendees.isEmpty()) {
+    List<TimeRange> optionalMeetingTimes = copyListTimeRnage(meetingTimes);
+    optionalMeetingTimes = findSchedule(timeRangeToRangeSet(meetingTimes), events, optionalGuests, request);
+    if (!optionalMeetingTimes.isEmpty()) {
       return optionalMeetingTimes;
     }
+
+    // find most optional guests that can attend
+    if (optionalMeetingTimes.isEmpty()) {
+      int n = optionalGuests.size(); 
+      List<List<TimeRange>> subsetsOptionalTimes = new ArrayList<List<TimeRange>>();
+      for (int i = 0; i < (1<<n); i++) {   
+        // current subset 
+        List<String> subsetOptionalAttendees = new ArrayList<String>();
+        for (int j = 0; j < n; j++) {
+          if ((i & (1 << j)) > 0) {
+            subsetOptionalAttendees.add((String) optionalGuests.toArray()[j]);
+          }
+        }
+        List<TimeRange> subsetOptionalMeetingTime = copyListTimeRnage(meetingTimes);
+        subsetOptionalMeetingTime = findSchedule(timeRangeToRangeSet(meetingTimes), events, subsetOptionalAttendees, request);
+        if (!subsetOptionalMeetingTime.isEmpty()) {
+          subsetsOptionalTimes.add(subsetOptionalMeetingTime);
+        } else {
+          break;
+        }
+      } 
+      if (!subsetsOptionalTimes.isEmpty()) {
+        return subsetsOptionalTimes.get(subsetsOptionalTimes.size() - 1);
```

Comment:
It might be the case that multiple subsets of optional attendees of size `n` have potential meeting slots. I would expect the behavior of this method to return all such meeting slots, not just those of an arbitrary optional attendee subset that with the maximal amount of attendees.

Note: my streams challenge will be a bit cleaner if you keep the current behavior, but it can still handle this modification. I'd recommend starting off trying to emulate the current behavior and then build on top of that.